### PR TITLE
cephfs_mirror: correctly set top level dir permissions

### PR DESCRIPTION
**Pull request by mchangir**
Fixes: https://tracker.ceph.com/issues/58678
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mchangir** [comment]
[Teuthology job](http://pulpito.front.sepia.ceph.com/mchangir-2023-02-10_06:21:13-fs:mirror-wip-mchangir-cephfs_mirror-correctly-set-top-level-dir-permissions-distro-default-smithi/) with the new integration test.

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 1]
```diff
@@ -1167,6 +1167,22 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
...
+
+  // sync the mode of the remote dir_root with that of the local dir_root
+  r = ceph_statx(m_local_mount, dir_root.c_str(), &tstx, CEPH_STATX_MODE,
+		 AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+  if (r < 0) {
+    derr << ": failed to fetch stat for local dir root: "
+	 << cpp_strerror(r) << dendl;
+    return r;
+  }
+  r = ceph_fchmod(m_remote_mount, fh.r_fd_dir_root, tstx.stx_mode);
```
Comment (vshankar) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1179]: Why not do this when creating the directory structure in `PeerReplayer::try_lock_directory()`?

Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1179]: mostly because this has nothing to do with directory locking

Comment (vshankar) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1179]: True, but that's when the remote directory is created, and it makes sense to fix the perms at that point.

**Comment by mchangir** [comment]
[Teuthology Job](http://pulpito.front.sepia.ceph.com/mchangir-2023-03-04_01:09:44-fs:mirror-wip-mchangir-cephfs_mirror-correctly-set-top-level-dir-permissions-distro-default-smithi/): successful

File: src/tools/cephfs_mirror/PeerReplayer.cc (hunk 1 above)
Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: revised the code

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 2]
```diff
@@ -1499,8 +1521,13 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
         dout(5) << ": picked dir_root=" << *dir_root << dendl;
         int r = register_directory(*dir_root, replayer);
         if (r == 0) {
-          sync_snaps(*dir_root, locker);
-          unregister_directory(*dir_root);
+	  r = sync_perms(*dir_root);
+	  if (r < 0) {
+	    _inc_failed_count(*dir_root);
```
Comment (dparmar18) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1526]: Maybe adding a dout/derr can be beneficial?
```
derr << ": failed to sync perms for dir_root=" << dir_root << dendl;
```
or
```
dout(10) << ": failed to sync perms for dir_root=" << dir_root << dendl;
```

Comment (dparmar18) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1526]: my bad, it's already done in `PeerReplayer::sync_perms`

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 3]
```diff
@@ -1167,6 +1188,7 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
   }
 
   struct ceph_statx tstx;
+
```
Comment (dparmar18) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1191]: is this intentional?

File: qa/tasks/cephfs/test_mirroring.py [hunk 4]
```diff
@@ -1261,3 +1261,38 @@ def test_cephfs_mirror_cancel_mirroring_and_readd(self):
...
         self.disable_mirroring(self.primary_fs_name, self.primary_fs_id)
+
+    def test_local_and_remote_dir_root_mode(self):
+        log.debug('reconfigure client auth caps')
+        self.mds_cluster.mon_manager.raw_cluster_cmd_result(
+            'auth', 'caps', "client.{0}".format(self.mount_b.client_id),
+                'mds', 'allow rw',
+                'mon', 'allow r',
+                'osd', 'allow rw pool={0}, allow rw pool={1}'.format(
+                    self.backup_fs.get_data_pool_name(), self.backup_fs.get_data_pool_name()))
```
Comment (dparmar18) [inline qa/tasks/cephfs/test_mirroring.py:1272]: f-strings would be better readable here

File: src/tools/cephfs_mirror/PeerReplayer.cc (hunk 3 above)
Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: deleted empty line

File: qa/tasks/cephfs/test_mirroring.py (hunk 4 above)
Comment (mchangir) [reply qa/tasks/cephfs/test_mirroring.py]: replaced with `f""` strings

**Comment by mchangir** [comment]
jenkins test make check arm64

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 5]
```diff
@@ -1128,6 +1128,27 @@ int PeerReplayer::pre_sync_check_and_open_handles(
   return 0;
 }
 
+int PeerReplayer::sync_perms(const std::string& path) {
+  int r = 0;
+  struct ceph_statx tstx;
+
+  // sync the mode of the remote dir_root with that of the local dir_root
+  r = ceph_statx(m_local_mount, path.c_str(), &tstx, CEPH_STATX_MODE,
+		 AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
```
Comment (dparmar18) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1137]: the mode sync happens at line 1143, the code on 1136 just tries to fetch the stat for the local path, maybe the comment on line 1135 be updated accordingly? or shift the comment to line 1143?

Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1137]: The comment is not at all necessary. However, I decided to keep it there for general guidance.

Comment (dparmar18) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1137]: then it can be re-positioned to the top maybe

Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: moved comment as the method description

File: qa/tasks/cephfs/test_mirroring.py [hunk 6]
```diff
@@ -1261,3 +1261,39 @@ def test_cephfs_mirror_cancel_mirroring_and_readd(self):
...
+
+        time.sleep(60)
+        self.check_peer_status(self.primary_fs_name, self.primary_fs_id,
+                               "client.mirror_remote@ceph", '/l1', 'snap0', 1)
+
+        mode_local = self.mount_a.run_shell(["stat", "--format=%A", "l1"]).stdout.getvalue().strip()
+        mode_remote = self.mount_b.run_shell(["stat", "--format=%A", "l1"]).stdout.getvalue().strip()
+        log.debug(f"local mode: {mode_local}, remote_mode: {mode_remote}")
+
+        self.assertTrue(mode_local == mode_remote)
```
[the remark this answers is not in the document]
Comment (mchangir) [reply qa/tasks/cephfs/test_mirroring.py]: done

**Review by dparmar18 — APPROVED** [review]
LGTM

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230309.101043

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#13-Mar-2023

**Pull request merged**